apps/api-runtime/src/main/java/com/akto/dependency/DependencyAnalyser.java

```diff
@@ -0,0 +1,272 @@
+                        Filters.eq(DependencyNode.API_COLLECTION_ID_RESP, apiCollectionIdResp),
+                        Filters.eq(DependencyNode.URL_RESP, urlResp),
+                        Filters.eq(DependencyNode.METHOD_RESP, methodResp),
+                        Filters.not(Filters.elemMatch(DependencyNode.PARAM_INFOS,
+                                Filters.and(
+                                        Filters.eq(DependencyNode.ParamInfo.REQUEST_PARAM, paramInfo.getRequestParam()),
+                                        Filters.eq(DependencyNode.ParamInfo.RESPONSE_PARAM, paramInfo.getResponseParam())
+                                )
+                        ))
+                );
+
+                Bson update2 = Updates.push(DependencyNode.PARAM_INFOS,
+                        new BasicDBObject(DependencyNode.ParamInfo.REQUEST_PARAM, paramInfo.getRequestParam())
+                                .append(DependencyNode.ParamInfo.RESPONSE_PARAM, paramInfo.getResponseParam())
+                                .append(DependencyNode.ParamInfo.COUNT, 0)
+                );
+
+                // this update is to add paramInfo if it doesn't exist. If exists nothing happens
+                UpdateOneModel<DependencyNode> updateOneModel2 = new UpdateOneModel<>(
+                        filter2, update2, new UpdateOptions().upsert(false)
+                );
+
+                Bson filter3 = Filters.and(
+                        Filters.eq(DependencyNode.API_COLLECTION_ID_REQ, apiCollectionIdReq),
+                        Filters.eq(DependencyNode.URL_REQ, urlReq),
+                        Filters.eq(DependencyNode.METHOD_REQ, methodReq),
+                        Filters.eq(DependencyNode.API_COLLECTION_ID_RESP, apiCollectionIdResp),
+                        Filters.eq(DependencyNode.URL_RESP, urlResp),
+                        Filters.eq(DependencyNode.METHOD_RESP, methodResp),
+                        Filters.eq(DependencyNode.PARAM_INFOS + "." + DependencyNode.ParamInfo.REQUEST_PARAM, paramInfo.getRequestParam()),
+                        Filters.eq(DependencyNode.PARAM_INFOS + "." + DependencyNode.ParamInfo.RESPONSE_PARAM, paramInfo.getResponseParam())
+                );
+
+                Bson update3 = Updates.inc(DependencyNode.PARAM_INFOS + ".$." + DependencyNode.ParamInfo.COUNT, paramInfo.getCount());
+
+                // this update runs everytime to update the count
+                UpdateOneModel<DependencyNode> updateOneModel3 = new UpdateOneModel<>(
+                        filter3, update3,  new UpdateOptions().upsert(false)
+                );
+
+                bulkUpdates.add(updateOneModel1);
+                bulkUpdates.add(updateOneModel2);
+                bulkUpdates.add(updateOneModel3);
+
+            }
+        }
+
+        // ordered has to be true or else won't work
+        if (bulkUpdates.size() > 0) DependencyNodeDao.instance.getMCollection().bulkWrite(bulkUpdates, new BulkWriteOptions().ordered(true));
```

Comment:
if there are n nodes, instead of ordered, can we insert findoneandupdate for 3 queries for a doc(which are ordered) in bulklist, and run unordered bulkupdate on this list of size n?